Show simple error dialog

cheld opened the pull request:
I have created a simple error dialog. It is a pop-up, because GCE also uses pop-ups. At this point, I have integrated the error dialog only at one place.

I would like to start with something simple and improve on it. I tried to encapsulate the error handling. 


Potential next steps:
- Improve errors from the backend, e.g. add error codes
- Improve/change presentation of errors
- resource handling seems to be similar in many cases, kind of duplicate code. Maybe we can improve here, too

cheld commented on src/app/frontend/deploy/createnamespace_controller.js:
```diff
@@ -35,6 +36,8 @@ export default class NamespaceDialogController {
     /** @private {!angular.$resource} */
     this.resource_ = $resource;
 
+    this.errorDialog_ = errorDialog;
```
JSDoc missing - I just noticed

bryk replied:
You can add type info: `@type {!./../../relative/path.FooClass}`.

bryk commented on src/app/frontend/chrome/errordialog_service.js:
```diff
@@ -0,0 +1,45 @@
+* Error dialog to display a errors that occured in the backend
+* @final
+**/
+export class ErrorDialog {
```
I think you should just create a class without the factory function. The reason is that the factory function does nothing, besides being an additional layer. Then use `.service` instead of `.factory` in the module. What do you think?

cheld replied:
unfortunately, lint complains about usage of .service

bryk commented on src/app/frontend/chrome/errordialog_service.js:
```diff
@@ -0,0 +1,45 @@
+* @final
+**/
+export class ErrorDialog {
+  constructor($mdDialog) { this.mdDialog_ = $mdDialog; }
```
Add `@type` information here :)

cheld replied:
Done.

bryk commented on src/app/frontend/chrome/errordialog_service.js:
```diff
@@ -0,0 +1,45 @@
+   * @param {string} message
+   * @export
+   */
+  open(message) {
```
This actually does not need to be `@exported` unless it is used in templates.

cheld replied:
It seems that export is required for Google Closure Compiler when using ErrorDialog as type

bryk commented on src/app/frontend/chrome/errordialog_service.js:
```diff
@@ -0,0 +1,45 @@
+      title: message,
+      ok: 'Close',
+    });
+    this.mdDialog_.show(alert).finally(() => { alert = undefined; });
```
Why is this alert = undefined needed?

bryk commented on src/app/frontend/chrome/errordialog_service.js:
```diff
@@ -0,0 +1,45 @@
+}
+
+/**
+* Error dialog to display a errors that occured in the backend
```
a errors -> errors

that occured in the backend -> I don't see a reason why we should limit this only to backend errors. I'd say it should display application errors or errors that we cannot recover from.

bryk commented on src/app/frontend/deploy/createnamespace_controller.js:
```diff
@@ -74,8 +77,9 @@ export default class NamespaceDialogController {
         (err) => {
-          this.log_.info('Error creating namespace:', err);
           this.mdDialog_.hide();
+          this.errorDialog_.open('Error creating namespace', err);
```
This is why we need type information. You called errorDialog_.open with second parameter, but the method does not accept it. With type information compiler would complain :)

bryk commented on src/app/frontend/chrome/errordialog_service.js:
```diff
@@ -0,0 +1,45 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Does this belong to chrome/ folder? Can we put it somewhere else?

cheld commented on src/app/frontend/deploy/createnamespace_controller.js:
```diff
@@ -22,10 +22,11 @@ export default class NamespaceDialogController {
    * @param {!md.$dialog} $mdDialog
    * @param {!angular.$log} $log
    * @param {!angular.$resource} $resource
+   * @param {!Object} errorDialog
```
this type causes lint error 

{!./../chrome/errordialog_service.ErrorDialog}

bryk replied:
What does the linter say?

cheld replied:
valid-jsdoc

JSDoc syntax error

bryk replied:
Disable valid-jsdoc rule then.

cheld replied:
{!./....../errordialog_service.ErrorDialog} is a default export. So it is still not working.

See also: https://github.com/kubernetes/dashboard/issues/159

Currently, we use default export a lot and I thought I should not change this in one place. If we change, we should change everywhere, as suggested by bryk.

bryk replied:
Add a todo with reference to the issue. We will fix it in a separate track. We just need to make it easy to track places to fix.

cheld replied:
Done

cheld commented on src/app/frontend/commons/errordialog_service.js:
```diff
@@ -0,0 +1,46 @@
+*
+* @final
+**/
+export class ErrorDialog {
```
Google Closure compiler complains if I do not export

bryk replied:
What's the error?

cheld replied:
src/app/frontend/deploy/createnamespace_controller.js:25: ERROR - Bad type annotation. Unknown type module$commons$errordialog_service.ErrorDialog
   * @param {!./../commons/errordialog_service.ErrorDialog} errorDialog
              ^

src/app/frontend/deploy/createnamespace_controller.js:39: ERROR - Bad type annotation. Unknown type module$commons$errordialog_service.ErrorDialog
    /** @private {!./../commons/errordialog_service.ErrorDialog} */
                   ^

2 error(s), 0 warning(s)

bryk replied:
Oh, yes. This needs to be exported if you want to use/reference it somewhere else. That's the idea. Exported things can be used/references in other files. Everything else is private.

cheld commented on src/app/frontend/commons/commons_module.js:
```diff
@@ -0,0 +1,11 @@
+                          [
+                            'ngMaterial',
+                          ])
+    .factory('errorDialog', errorDialogService);
```
.service causes lint error

bryk replied:
What does the linter say?

cheld replied:
angular/no-service-method

You should prefer the factory(9 method instead of service()

bryk replied:
I see. You can safely disable this lint check. I checked the rule and it does not apply to our codebase: https://github.com/johnpapa/angular-styleguide#style-y040

We're using ES6 classes everywhere so the lint rule adds confusion rather than any benefits.

cheld commented:
PTAL
 I can't do better ;-(

bryk commented on src/test/frontend/deploy/createnamespace_controller_test.js:
```diff
@@ -22,7 +22,7 @@ describe('Deploy controller', () => {
 
     angular.mock.inject(($mdDialog, $resource, $log) => {
-      ctrl = new NamespaceDialogController($mdDialog, $resource, $log, []);
+      ctrl = new NamespaceDialogController($mdDialog, $resource, $log, null, []);
```
If youre already touching this. Use $controller service to create controllers, its much easier than manual work. See angular documentation for more info.

bryk commented on src/app/frontend/index_module.js:
```diff
@@ -35,6 +36,7 @@ export default angular.module(
                             'ngSanitize',
                             'ui.router',
                             chromeModule.name,
+                            commonsModule.name,
```
Index does not depend on commons module, deploy module does. Dependencies should be declared in the places where they are actually used. Otherwise we would end up with one catch-all module listing all dependencies, which voids the idea of reusability.

cheld commented:
PTAL

I added the error messages

bryk commented on src/app/frontend/commons/commons_module.js:
```diff
@@ -0,0 +1,11 @@
+import errorDialogService from './errordialog_service';
+
```
I'm a little bit worried to introduce `commons` folder. This is because it is like failing to organize codebase and whenever it is hard to find a place for a thing, it lands in common. As a result, you actually dont know what to expect in this folder.

Do you have any idea on what's the contract for this folder? What to place here? I was thinking that such folder can be for non-functional and application-wide services and UI components. 

.. still not convinced and on the fence ...

cheld replied:
I agree. What about making it similar to angular-seed project

.../components/errorhandling
.../components/i18n

more structure and it is not too easy to just dump code into a central place

bryk replied:
This looks good to me. Angular 2.0 and 1.5 also has strong connection with components/ directory so this would align us a little with it.

cheld replied:
I found a common folder after rebase. So, I moved the code to this folder. Do you agree?

cheld commented:
@floreks CC

There are two problems left:
1. @param {!./../commons/errordialog_service.ErrorDialog}  causes JSDoc validation error
2. where to place the code

any ideas?

cheld commented:
@floreks @maciaszczykm CC

PTAL

2 issues open. See comments

bryk commented on src/test/frontend/deploy/createnamespace_controller_test.js:
```diff
@@ -14,16 +14,17 @@
 
   beforeEach(() => {
     angular.mock.module(deployModule.name);
+    angular.mock.module(commonsModule.name);
```
You dont need this here. Deploy module already depends on it.

bryk commented:
@floreks @maciaszczykm Are you reviewing this?

maciaszczykm commented on src/app/frontend/chrome/chrome_module.js:
```diff
@@ -13,7 +13,6 @@
 // limitations under the License.
 
 import chromeDirective from './chrome_directive';
-
```
Should it be removed?

bryk replied:
Yes, I'd say revert.

maciaszczykm commented:
@bryk I already had reviewed it internally some time ago and still I cannot find any solution for problem with `{!./../chrome/errordialog_service.ErrorDialog}`. Only solution which works for me is using `{!Object}`, but I'm afraid we cannot leave it like that.

Message should be more specific too, but I believe it will be improved in the future:

![Test](http://i.imgur.com/bZB9Sxq.png)

Everything else LGTM.

bryk commented:
> I already had reviewed it internally some time ago and still I didn't find any solution for problem with {!./../chrome/errordialog_service.ErrorDialog}. Only solution which works for me is using {!Object}, but I'm afraid we cannot leave it like that.

I say we should disable valid-json eslint rule for now. It doesn't add that much value anyway.

bryk commented on src/app/frontend/common/errorhandling/errordialog_service.js:
```diff
@@ -0,0 +1,36 @@
+// limitations under the License.
+
+/**
+* Error dialog to display errors which could not be responded properly
```
This comment block has invalid formatting. Needs space.

bryk commented on src/app/frontend/common/errorhandling/errordialog_service.js:
```diff
@@ -0,0 +1,36 @@
+* Error dialog to display errors which could not be responded properly
+*
+* @final
+**/
```
Close JsDoc with `*/`.

bryk commented on src/app/frontend/common/errorhandling/errordialog_service.js:
```diff
@@ -0,0 +1,36 @@
+* @final
+**/
+export default class ErrorDialog {
+  constructor($mdDialog) { this.mdDialog_ = $mdDialog; }
```
Add `@ngInject` annotations and annotate `this.mdDialog_`

bryk commented:
Looks nice. A few small comments and let's merge.

cheld commented:
I will write PTAL when I have addressed all comments

cheld commented on src/app/frontend/deploy/createnamespace.html:
```diff
@@ -19,7 +19,7 @@
     <p>The new namespace will be added to the cluster.</p>
     <!-- TODO(maciaszczykm): Missing detailed cluster info. -->
-    <form name="namespaceForm" ng-submit="ctrl.createNamespace()"
+    <form name="ctrl.namespaceForm" ng-submit="ctrl.createNamespace()"
```
The validation did not work any longer. Not sure why. I fixed it by adding ctrl prefix....

cheld commented:
PTAL

Sorry for the long delay, but I was on holidays and found no time to finish up. I think I have addressed all your comments.

After merge, I add a comment to issue 159 about fixing object type.

bryk commented:
LGTM. Will do final check tomorrow and then merge :)

bryk commented on src/app/frontend/common/errorhandling/errordialog_service.js:
```diff
@@ -0,0 +1,38 @@
+ * @final
+ **/
+export default class ErrorDialog {
+  constructor($mdDialog) {
```
Hey, this needs `@ngInject` annotation.

bryk commented on src/app/frontend/common/errorhandling/errordialog_service.js:
```diff
@@ -0,0 +1,38 @@
+ * Error dialog to display errors which could not be responded properly
+ *
+ * @final
+ **/
```
Close this with `*/` to have valid JsDoc :)

bryk commented:
This does not work on `serve:prod`. You need to add `@ngInject` in commented place. Then we can merge :)

cheld commented:
Oh my gosh! Usually, I do test serve:prod, but not this time. 

PTAL

bryk commented:
:) Merging
